presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java

```diff
@@ -0,0 +1,436 @@
+                    node.getConfidence(),
+                    Optional.<Symbol>empty());
+        }
+
+        @Override
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<AggregateInfo>> context)
+        {
+            Optional<AggregateInfo> aggregateInfo = context.get();
+            // presence of aggregateInfo => mask also present
+            if (aggregateInfo.isPresent() && aggregateInfo.get().getMask().get().equals(node.getMarkerSymbol())) {
+                PlanNode source = node.getSource();
+
+                List<Symbol> allSymbols = new ArrayList<>();
+                List<Symbol> groupBySymbols = aggregateInfo.get().getGroupBySymbols(); // a
+                allSymbols.addAll(groupBySymbols);
+
+                List<Symbol> nonDistinctAggregateSymbols = aggregateInfo.get().getOriginalNonDistinctAggregateArgs(); //b
+                allSymbols.addAll(nonDistinctAggregateSymbols);
+
+                Symbol distinctSymbol = Iterables.getOnlyElement(aggregateInfo.get().getOriginalDistinctAggregateArgs()); // c
+                Symbol distinctSymbolForNonDistinctAggregates = distinctSymbol;
+                allSymbols.add(distinctSymbol);
+
+                // If same symbol in distinct and non-distinct aggregations
+                if (nonDistinctAggregateSymbols.contains(distinctSymbol)) {
+                    Symbol newSymbol = symbolAllocator.newSymbol(distinctSymbol.getName(), symbolAllocator.getTypes().get(distinctSymbol));
+                    nonDistinctAggregateSymbols.set(nonDistinctAggregateSymbols.indexOf(distinctSymbol), newSymbol);
+                    distinctSymbolForNonDistinctAggregates = newSymbol;
+
+                    // Add project node to get stream for new symbol
+                    ImmutableMap.Builder builder = new ImmutableMap.Builder();
+                    for (Symbol symbol : node.getSource().getOutputSymbols()) {
+                        builder.put(symbol, symbol.toSymbolReference());
+                    }
+                    // Cast to avoid UnaliasSymbolReference from removing this projection
+                    // TODO: find better way to prevent UnaliaSymbolRefrence optimizing out the project
+                    builder.put(newSymbol, new Cast(distinctSymbol.toSymbolReference(),
+                            symbolAllocator.getTypes().get(distinctSymbol).getDisplayName()));
+
+                    source = new ProjectNode(idAllocator.getNextId(), node.getSource(), builder.build());
+                }
+
+                Symbol groupSymbol = symbolAllocator.newSymbol("group", BigintType.BIGINT); // g
+
+                List<List<Symbol>> groups = new ArrayList<>();
+                // g0 = {allGBY_Symbols + allNonDistinctAggregateSymbols}
+                // g1 = {allGBY_Symbols + Distinct Symbol}
+                // symbols present in Group_i will be set, rest will be Null
+
+                //g0
+                if (!nonDistinctAggregateSymbols.isEmpty()) {
```

Comment:
You are right, wont happen right now. This is there because I was thinking of extending this to multiple distinct aggregations. Will remove it for now.